src/Database/Esqueleto/Internal/Internal.hs

```diff
@@ -1406,25 +1407,42 @@ data Insertion a
 -- are no guarantees that they will behave the same.
 --
 -- @since 2.2.7
-data LockingKind
-    = ForUpdate
+data LockingKind where
+    ForUpdate :: LockingKind
       -- ^ @FOR UPDATE@ syntax.  Supported by MySQL, Oracle and
       -- PostgreSQL.
       --
       -- @since 2.2.7
-    | ForUpdateSkipLocked
+    ForUpdateSkipLocked :: LockingKind
       -- ^ @FOR UPDATE SKIP LOCKED@ syntax.  Supported by MySQL, Oracle and
       -- PostgreSQL.
       --
       -- @since 2.2.7
-    | ForShare
+    ForShare :: LockingKind
       -- ^ @FOR SHARE@ syntax.  Supported by PostgreSQL.
       --
       -- @since 2.2.7
-    | LockInShareMode
+    LockInShareMode :: LockingKind
       -- ^ @LOCK IN SHARE MODE@ syntax.  Supported by MySQL.
       --
       -- @since 2.2.7
+    ForUpdateOfSkipLocked :: [LockableEntity] -> LockingKind
+      -- ^ @FOR UPDATE OF tablename SKIP LOCKED@ syntax. Supported by MySQL, and PostgreSQL
+      --
+      -- @since 3.5.4.2
+
+-- | Wraps table type entities for usage in LockingKind. Usage:
+-- Example of usage:
+--
+-- @
+-- select $ do
+--   from $ \(p `InnerJoin` b) -> do
+--     on (p ^. PersonId ==. b ^. BlogPostAuthorId)
+--   locking (ForUpdateOfSkipLocked [LockableEntity p,LockableEntity b])
+--   return p
+-- @
+data LockableEntity where
+  LockableEntity :: PersistEntity val => (SqlExpr (Entity val)) -> LockableEntity
```

Comment:
I think we actually don't want `(SqlExpr (Maybe (Entity val))` to be available for the locking clause because of the error:

```
FOR UPDATE cannot be applied to the nullable side of an outer join
```
I'm not sure there's another case where we want to be able to do this with a nullable side?

Semantically `:&` is meant to represent "A left-precedence pair. Pronounced \"and\". Used to represent expressions
that have been joined together." so I'm a bit hesitant to use it for this purpose but maybe it's fine since the main use case is on joins?  I'm also not quite sure how to have an arbitrary number of tables with the tuple solution as well. 

For those reasons I do overall prefer a list since I think it lines up more closely with the syntax but don't feel super strongly about this so I'm happy to use this pattern instead if you strongly prefer it.